lib/elixir/lib/map_set.ex

```diff
@@ -149,16 +200,24 @@ defmodule MapSet do
   @spec intersection(t, t) :: t
   def intersection(%MapSet{map: map1}, %MapSet{map: map2}) do
     {map1, map2} = order_by_size(map1, map2)
-    map = :maps.fold(fn value, _, acc ->
-      if Map.has_key?(map2, value) do
-        Map.put(acc, value, true)
-      else
-        acc
-      end
-    end, %{}, map1)
+
+    map = map1
+    |> Map.keys
+    |> acc_intersection(map2, [])
+
     %MapSet{map: map}
   end
 
+  defp acc_intersection([], _map2, acc), do: :maps.from_list(acc)
+  defp acc_intersection([k | rest], map2, acc) do
+    acc = if Map.has_key?(map2, k) do
+      [{k, true} | acc]
+    else
+      acc
+    end
+    acc_intersection(rest, map2, acc)
+  end
```

Comment:
Heh yeah this is literally Map.take with manual recursion rather than Enum.reduce. Ironic, given that I wrote that version of Map.take last time I did this :S.

I'll make these optimizations to Map.take, and just make this Map.take.